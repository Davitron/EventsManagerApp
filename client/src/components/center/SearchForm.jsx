import React, { Component } from 'react';
import { Form } from 'semantic-ui-react';
import PropTypes from 'prop-types';

const facilities = [
  { key: '1', text: 'CCTV', value: 'cctv' },
  { key: '2', text: 'Vip Lounge', value: 'vip lounge' },
  { key: '3', text: 'Projector', value: 'projector' },
  { key: '4', text: 'Security', value: 'security' },
  { key: '5', text: 'WIFI', value: 'wifi' }
];

/**
 * @class
 */
class SearchForm extends Component {
  /**
   *
   * @param {*} props
   */
  constructor(props) {
    super(props);
    this.state = {
      query: {},
      states: []
    };

    this.onSubmit = this.onSubmit.bind(this);
    this.onChange = this.onChange.bind(this);
    this.renderLabel = this.renderLabel.bind(this);
  }


  /**
   *
   * @param {*} nextProps
   *
   * @returns {void}
   */
  componentWillReceiveProps(nextProps) {
    if (nextProps.states !== this.props.states && nextProps.states.length > 0) {
      const options = nextProps.states.map((state) => {
        const newObj = {};
        newObj.key = state.id;
        newObj.text = state.stateName;
        newObj.value = state.id;
        return newObj;
      });
      options.unshift({ key: 'all', text: 'States', value: null });
      this.setState({ states: options });
    }
  }

  /**
   *
  *
  * @param {object} event
  *
  * @param {object} data
  *
  * @returns {void}
  *
  * this handles the event when any property in the state changes
  */
  onChange(event, data) {
    const { value, name } = data;
    const { query } = this.state;
    this.setState({
      query: {
        ...query,
        [name]: value
      }
    }, () => {
      document.getElementById('search').focus();
    });
  }

  /**
  * @param {object} event
  *
  * @returns {void}
  *
  * this handles the event when form is submitted
  */
  onSubmit(event) {
    event.preventDefault();
    const { query } = this.state;
    if (query.state === null) delete query.state;
    this.props.onSearch(query);
  }

  /**
   *
   * @param {*} label
   *
   * @returns {object} selected
   */
  renderLabel(label) {
    return {
      color: 'blue',
      content: label.text,
      icon: 'check',
    };
  }

  /**
   * @returns {view} - html
   */
  render() {
    const { states } = this.state;
    return (
      <Form onSubmit={this.onSubmit}>
        <Form.Group widths="equal">
          <Form.Input id="search" name="search" fluid placeholder="Name or Addres" onChange={this.onChange} />
          <Form.Select
            className="states"
            options={states}
            search
            name="state"
            fluid
            placeholder="State"
            onChange={this.onChange}
          />
          <Form.Input type="number" min="50" name="capacity" fluid placeholder="Capacity" onChange={this.onChange} />
          <Form.Dropdown
            multiple
            search
            selection
            name="facilities"
            fluid
            options={facilities}
            placeholder="Choose Facilities"
            renderLabel={this.renderLabel}
            onChange={this.onChange}
          />
          <Form.Button primary fluid onClick={this.onSubmit}>Search</Form.Button>
        </Form.Group>
      </Form>
    );
  }
}

SearchForm.propTypes = {
  onSearch: PropTypes.func.isRequired,
  states: PropTypes.arrayOf(() => null)
};

SearchForm.defaultProps = {
  states: []
};

export default SearchForm;
