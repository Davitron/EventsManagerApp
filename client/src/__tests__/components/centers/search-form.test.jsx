import React from 'react';
import expect from 'expect';
import { shallow } from 'enzyme';
import SearchForm from '../../../components/center/SearchForm';

let mountedComponent;
let wrapper;

let props;

let onSearch = event => event;
onSearch = jest.fn();

// const document = {
//   getElementById: (id) => { id, focus: jest.fn() }
// };


const getComponent = () => {
  if (!mountedComponent) {
    props = {
      onSearch
    };
    mountedComponent = shallow(<SearchForm {...props} />);
  }
  return mountedComponent;
};

describe('Search Component', () => {
  beforeAll(() => {
    wrapper = getComponent();
  });

  it('component successfully rendered', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('testing component will recieve props', () => {
    const nextProps = {
      states: [
        { id: 1, stateName: 'Abia' },
        { id: 2, stateName: 'Ogun' },
        { id: 3, stateName: 'FCT' }
      ]
    };

    wrapper.instance().componentWillReceiveProps(nextProps);
    expect(wrapper.state().states.length).toEqual(nextProps.states.length + 1);
  });

  it('testing on change function', () => {
    const data = { name: 'search', value: 'test' };
    const event = { name: 'search', value: 'test' };
    wrapper.instance().onChange(event, data);
    expect(wrapper.state().query[data.name]).toEqual(data.value);
  });

  it('testing on onSubmit function', () => {
    wrapper = getComponent();
    const event = { preventDefault: () => undefined };
    event.preventDefault = jest.fn();
    wrapper.setState({
      query: {
        state: null
      }
    });
    wrapper.instance().onSubmit(event);
    expect(event.preventDefault).toHaveBeenCalled();
    expect(onSearch).toHaveBeenCalled();
  });

  it('testing on renderLabel function', () => {
    const event = { preventDefault: () => undefined };
    event.preventDefault = jest.fn();
    wrapper.instance().onSubmit(event);
    expect(event.preventDefault).toHaveBeenCalled();
    expect(onSearch).toHaveBeenCalled();
  });

  it('testing renderLabel function', () => {
    const param = { key: 1, text: 'hello' };
    const label = wrapper.instance().renderLabel(param);
    expect(label.color).toEqual('blue');
    expect(label.content).toEqual(param.text);
    expect(label.icon).toEqual('check');
  });
});
