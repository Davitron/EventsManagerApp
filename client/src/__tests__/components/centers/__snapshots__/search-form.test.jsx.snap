// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Component component successfully rendered 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SearchForm
    onSearch={[MockFunction]}
    states={Array []}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "as": "form",
      "children": <FormGroup
        widths="equal"
      >
        <FormInput
          as={[Function]}
          control={[Function]}
          fluid={true}
          id="search"
          name="search"
          onChange={[Function]}
          placeholder="Name or Addres"
        />
        <FormSelect
          as={[Function]}
          className="states"
          control={[Function]}
          fluid={true}
          name="state"
          onChange={[Function]}
          options={Array []}
          placeholder="State"
          search={true}
        />
        <FormInput
          as={[Function]}
          control={[Function]}
          fluid={true}
          min="50"
          name="capacity"
          onChange={[Function]}
          placeholder="Capacity"
          type="number"
        />
        <FormDropdown
          as={[Function]}
          control={[Function]}
          fluid={true}
          multiple={true}
          name="facilities"
          onChange={[Function]}
          options={
            Array [
              Object {
                "key": "1",
                "text": "CCTV",
                "value": "cctv",
              },
              Object {
                "key": "2",
                "text": "Vip Lounge",
                "value": "vip lounge",
              },
              Object {
                "key": "3",
                "text": "Projector",
                "value": "projector",
              },
              Object {
                "key": "4",
                "text": "Security",
                "value": "security",
              },
              Object {
                "key": "5",
                "text": "WIFI",
                "value": "wifi",
              },
            ]
          }
          placeholder="Choose Facilities"
          renderLabel={[Function]}
          search={true}
          selection={true}
        />
        <FormButton
          as={[Function]}
          control={[Function]}
          fluid={true}
          onClick={[Function]}
          primary={true}
        >
          Search
        </FormButton>
      </FormGroup>,
      "onSubmit": [Function],
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <FormInput
            as={[Function]}
            control={[Function]}
            fluid={true}
            id="search"
            name="search"
            onChange={[Function]}
            placeholder="Name or Addres"
          />,
          <FormSelect
            as={[Function]}
            className="states"
            control={[Function]}
            fluid={true}
            name="state"
            onChange={[Function]}
            options={Array []}
            placeholder="State"
            search={true}
          />,
          <FormInput
            as={[Function]}
            control={[Function]}
            fluid={true}
            min="50"
            name="capacity"
            onChange={[Function]}
            placeholder="Capacity"
            type="number"
          />,
          <FormDropdown
            as={[Function]}
            control={[Function]}
            fluid={true}
            multiple={true}
            name="facilities"
            onChange={[Function]}
            options={
              Array [
                Object {
                  "key": "1",
                  "text": "CCTV",
                  "value": "cctv",
                },
                Object {
                  "key": "2",
                  "text": "Vip Lounge",
                  "value": "vip lounge",
                },
                Object {
                  "key": "3",
                  "text": "Projector",
                  "value": "projector",
                },
                Object {
                  "key": "4",
                  "text": "Security",
                  "value": "security",
                },
                Object {
                  "key": "5",
                  "text": "WIFI",
                  "value": "wifi",
                },
              ]
            }
            placeholder="Choose Facilities"
            renderLabel={[Function]}
            search={true}
            selection={true}
          />,
          <FormButton
            as={[Function]}
            control={[Function]}
            fluid={true}
            onClick={[Function]}
            primary={true}
          >
            Search
          </FormButton>,
        ],
        "widths": "equal",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "as": [Function],
            "control": [Function],
            "fluid": true,
            "id": "search",
            "name": "search",
            "onChange": [Function],
            "placeholder": "Name or Addres",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "as": [Function],
            "className": "states",
            "control": [Function],
            "fluid": true,
            "name": "state",
            "onChange": [Function],
            "options": Array [],
            "placeholder": "State",
            "search": true,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "as": [Function],
            "control": [Function],
            "fluid": true,
            "min": "50",
            "name": "capacity",
            "onChange": [Function],
            "placeholder": "Capacity",
            "type": "number",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "as": [Function],
            "control": [Function],
            "fluid": true,
            "multiple": true,
            "name": "facilities",
            "onChange": [Function],
            "options": Array [
              Object {
                "key": "1",
                "text": "CCTV",
                "value": "cctv",
              },
              Object {
                "key": "2",
                "text": "Vip Lounge",
                "value": "vip lounge",
              },
              Object {
                "key": "3",
                "text": "Projector",
                "value": "projector",
              },
              Object {
                "key": "4",
                "text": "Security",
                "value": "security",
              },
              Object {
                "key": "5",
                "text": "WIFI",
                "value": "wifi",
              },
            ],
            "placeholder": "Choose Facilities",
            "renderLabel": [Function],
            "search": true,
            "selection": true,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "as": [Function],
            "children": "Search",
            "control": [Function],
            "fluid": true,
            "onClick": [Function],
            "primary": true,
          },
          "ref": null,
          "rendered": "Search",
          "type": [Function],
        },
      ],
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "as": "form",
        "children": <FormGroup
          widths="equal"
        >
          <FormInput
            as={[Function]}
            control={[Function]}
            fluid={true}
            id="search"
            name="search"
            onChange={[Function]}
            placeholder="Name or Addres"
          />
          <FormSelect
            as={[Function]}
            className="states"
            control={[Function]}
            fluid={true}
            name="state"
            onChange={[Function]}
            options={Array []}
            placeholder="State"
            search={true}
          />
          <FormInput
            as={[Function]}
            control={[Function]}
            fluid={true}
            min="50"
            name="capacity"
            onChange={[Function]}
            placeholder="Capacity"
            type="number"
          />
          <FormDropdown
            as={[Function]}
            control={[Function]}
            fluid={true}
            multiple={true}
            name="facilities"
            onChange={[Function]}
            options={
              Array [
                Object {
                  "key": "1",
                  "text": "CCTV",
                  "value": "cctv",
                },
                Object {
                  "key": "2",
                  "text": "Vip Lounge",
                  "value": "vip lounge",
                },
                Object {
                  "key": "3",
                  "text": "Projector",
                  "value": "projector",
                },
                Object {
                  "key": "4",
                  "text": "Security",
                  "value": "security",
                },
                Object {
                  "key": "5",
                  "text": "WIFI",
                  "value": "wifi",
                },
              ]
            }
            placeholder="Choose Facilities"
            renderLabel={[Function]}
            search={true}
            selection={true}
          />
          <FormButton
            as={[Function]}
            control={[Function]}
            fluid={true}
            onClick={[Function]}
            primary={true}
          >
            Search
          </FormButton>
        </FormGroup>,
        "onSubmit": [Function],
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <FormInput
              as={[Function]}
              control={[Function]}
              fluid={true}
              id="search"
              name="search"
              onChange={[Function]}
              placeholder="Name or Addres"
            />,
            <FormSelect
              as={[Function]}
              className="states"
              control={[Function]}
              fluid={true}
              name="state"
              onChange={[Function]}
              options={Array []}
              placeholder="State"
              search={true}
            />,
            <FormInput
              as={[Function]}
              control={[Function]}
              fluid={true}
              min="50"
              name="capacity"
              onChange={[Function]}
              placeholder="Capacity"
              type="number"
            />,
            <FormDropdown
              as={[Function]}
              control={[Function]}
              fluid={true}
              multiple={true}
              name="facilities"
              onChange={[Function]}
              options={
                Array [
                  Object {
                    "key": "1",
                    "text": "CCTV",
                    "value": "cctv",
                  },
                  Object {
                    "key": "2",
                    "text": "Vip Lounge",
                    "value": "vip lounge",
                  },
                  Object {
                    "key": "3",
                    "text": "Projector",
                    "value": "projector",
                  },
                  Object {
                    "key": "4",
                    "text": "Security",
                    "value": "security",
                  },
                  Object {
                    "key": "5",
                    "text": "WIFI",
                    "value": "wifi",
                  },
                ]
              }
              placeholder="Choose Facilities"
              renderLabel={[Function]}
              search={true}
              selection={true}
            />,
            <FormButton
              as={[Function]}
              control={[Function]}
              fluid={true}
              onClick={[Function]}
              primary={true}
            >
              Search
            </FormButton>,
          ],
          "widths": "equal",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "as": [Function],
              "control": [Function],
              "fluid": true,
              "id": "search",
              "name": "search",
              "onChange": [Function],
              "placeholder": "Name or Addres",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "as": [Function],
              "className": "states",
              "control": [Function],
              "fluid": true,
              "name": "state",
              "onChange": [Function],
              "options": Array [],
              "placeholder": "State",
              "search": true,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "as": [Function],
              "control": [Function],
              "fluid": true,
              "min": "50",
              "name": "capacity",
              "onChange": [Function],
              "placeholder": "Capacity",
              "type": "number",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "as": [Function],
              "control": [Function],
              "fluid": true,
              "multiple": true,
              "name": "facilities",
              "onChange": [Function],
              "options": Array [
                Object {
                  "key": "1",
                  "text": "CCTV",
                  "value": "cctv",
                },
                Object {
                  "key": "2",
                  "text": "Vip Lounge",
                  "value": "vip lounge",
                },
                Object {
                  "key": "3",
                  "text": "Projector",
                  "value": "projector",
                },
                Object {
                  "key": "4",
                  "text": "Security",
                  "value": "security",
                },
                Object {
                  "key": "5",
                  "text": "WIFI",
                  "value": "wifi",
                },
              ],
              "placeholder": "Choose Facilities",
              "renderLabel": [Function],
              "search": true,
              "selection": true,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "as": [Function],
              "children": "Search",
              "control": [Function],
              "fluid": true,
              "onClick": [Function],
              "primary": true,
            },
            "ref": null,
            "rendered": "Search",
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
